server/src/main/java/org/opensearch/repositories/blobstore/RemoteStoreShardCleanupTask.java

```diff
@@ -39,25 +36,16 @@ private static String indexShardIdentifier(String indexUUID, ShardId shardId) {
 
     @Override
     public void run() {
-        // TODO: this is the best effort at the moment since there is still a known race condition scenario in this
-        // method which needs to be handled where one of the thread just came out of while loop and removed the
-        // entry from ongoingRemoteDirectoryCleanup, and another thread added new pending task in the map.
-        // we need to introduce semaphores/locks to avoid that situation which introduces the overhead of lock object
-        // cleanups. however, there will be no scenario where two threads run cleanup for same shard at same time.
-        // <issue-link>
-        if (pendingRemoteDirectoryCleanups.put(shardIdentifier, task) == null) {
-            if (ongoingRemoteDirectoryCleanups.add(shardIdentifier)) {
-                while (pendingRemoteDirectoryCleanups.containsKey(shardIdentifier)) {
-                    Runnable newTask = pendingRemoteDirectoryCleanups.get(shardIdentifier);
-                    pendingRemoteDirectoryCleanups.remove(shardIdentifier);
-                    newTask.run();
-                }
+        // If there is already a same task ongoing for a shard, we need to skip the new task to avoid multiple
+        // concurrent cleanup of same shard.
+        if (ongoingRemoteDirectoryCleanups.add(shardIdentifier)) {
```

Comment:
Thanks. this will remove the leak issue, but lets still add the comment around best effort since all other tasks will not be taken up which can result in files not getting cleaned up.

Reply:
makes sense, updated the comment here: https://github.com/opensearch-project/OpenSearch/pull/12672/files#diff-0c75016719f0865438856949c6ccf7b3aa0cd121456bbe3826710a422537a736R1156